src/lib/db/account-store.ts

```diff
@@ -0,0 +1,208 @@
+/* eslint camelcase: "off" */
+
+import { Knex } from 'knex';
+import { Logger, LogProvider } from '../logger';
+import User from '../types/user';
+
+import NotFoundError from '../error/notfound-error';
+import {
+    ICreateUser,
+    IUserLookup,
+    IUserUpdateFields,
+} from '../types/stores/user-store';
+import { IAccountStore } from '../types';
+
+const TABLE = 'users';
+
+const USER_COLUMNS_PUBLIC = [
+    'id',
+    'name',
+    'username',
+    'email',
+    'image_url',
+    'seen_at',
+    'is_service',
+];
+
+const USER_COLUMNS = [...USER_COLUMNS_PUBLIC, 'login_attempts', 'created_at'];
+
+const emptify = (value) => {
+    if (!value) {
+        return undefined;
+    }
+    return value;
+};
+
+const safeToLower = (s?: string) => (s ? s.toLowerCase() : s);
+
+const mapUserToColumns = (user: ICreateUser) => ({
+    name: user.name,
+    username: user.username,
+    email: safeToLower(user.email),
+    image_url: user.imageUrl,
+});
+
+const rowToUser = (row) => {
+    if (!row) {
+        throw new NotFoundError('No user found');
+    }
+    return new User({
+        id: row.id,
+        name: emptify(row.name),
+        username: emptify(row.username),
+        email: emptify(row.email),
+        imageUrl: emptify(row.image_url),
+        loginAttempts: row.login_attempts,
+        seenAt: row.seen_at,
+        createdAt: row.created_at,
+        isService: row.is_service,
+    });
+};
+
+export class AccountStore implements IAccountStore {
+    private db: Knex;
+
+    private logger: Logger;
+
+    constructor(db: Knex, getLogger: LogProvider) {
+        this.db = db;
+        this.logger = getLogger('account-store.ts');
+    }
+
+    async update(id: number, fields: IUserUpdateFields): Promise<User> {
+        await this.activeAccounts()
+            .where('id', id)
+            .update(mapUserToColumns(fields));
+        return this.get(id);
+    }
+
+    async insert(user: ICreateUser): Promise<User> {
+        const rows = await this.db(TABLE)
+            .insert(mapUserToColumns(user))
+            .returning(USER_COLUMNS);
+        return rowToUser(rows[0]);
+    }
+
+    async upsert(user: ICreateUser): Promise<User> {
+        const id = await this.hasAccount(user);
+
+        if (id) {
+            return this.update(id, user);
+        }
+        return this.insert(user);
+    }
```

Comment:
If we remove the modification methods we can get rid of all this code

Reply:
See: https://github.com/Unleash/unleash/pull/2918#discussion_r1073607177